api/agent/protocol/json.go

```diff
@@ -37,132 +46,154 @@ type CallResponseHTTP struct {
 type jsonIn struct {
 	CallID      string          `json:"call_id"`
 	Deadline    string          `json:"deadline"`
-	Body        string          `json:"body"`
+	Body        json.RawMessage `json:"body"`
 	ContentType string          `json:"content_type"`
 	Protocol    CallRequestHTTP `json:"protocol"`
 }
 
 // jsonOut the expected response from the function container
 type jsonOut struct {
-	Body        string            `json:"body"`
+	Body        json.RawMessage   `json:"body"`
 	ContentType string            `json:"content_type"`
 	Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
 }
 
-// JSONProtocol converts stdin/stdout streams from HTTP into JSON format.
-type JSONProtocol struct {
-	// These are the container input streams, not the input from the request or the output for the response
-	in  io.Writer
-	out io.Reader
-}
-
 func (p *JSONProtocol) IsStreamable() bool {
 	return true
 }
 
-func (h *JSONProtocol) writeJSONToContainer(ci CallInfo) error {
-	buf := bufPool.Get().(*bytes.Buffer)
-	buf.Reset()
-	defer bufPool.Put(buf)
+func (h *JSONProtocol) writeJSONToContainer(ci *event.Event) (string, error) {
+
+	callID, err := ci.GetCallID()
+	if err != nil {
+		return "", err
+	}
 
-	_, err := io.Copy(buf, ci.Input())
+	deadline, err := ci.GetDeadline()
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	body := buf.String()
+	var method, requestURL string
+	var headers http.Header
+	var ext *httpevent.HTTPReqExt
+	if ci.HasExtension(httpevent.ExtIoFnProjectHTTPReq) {
+		err = ci.ReadExtension(httpevent.ExtIoFnProjectHTTPReq, &ext)
+		if err != nil {
+			fmt.Errorf("invalid HTTP metadata on incoming inputs: %s", err)
+		}
+		method = ext.Method
+		requestURL = ext.RequestURL
+		headers = ext.Headers
+	} else {
+		method = "GET"
+		requestURL = FakeSourceURL
+	}
+
+	if headers == nil {
+		headers = make(map[string][]string)
+	}
 
 	in := jsonIn{
-		Body:        body,
-		ContentType: ci.ContentType(),
-		CallID:      ci.CallID(),
-		Deadline:    ci.Deadline().String(),
+		Body:        ci.Data,
+		ContentType: ci.ContentType,
+		CallID:      callID,
+		Deadline:    deadline.String(),
 		Protocol: CallRequestHTTP{
-			Type:       ci.ProtocolType(),
-			Method:     ci.Method(),
-			RequestURL: ci.RequestURL(),
-			Headers:    ci.Headers(),
+			Type:       "http",
+			Method:     method,
+			RequestURL: requestURL,
+			Headers:    headers,
 		},
 	}
 
-	return json.NewEncoder(h.in).Encode(in)
+	err = json.NewEncoder(h.in).Encode(in)
+	if err != nil {
+		return "", err
+	}
+	return callID, nil
+
 }
 
-func (h *JSONProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) error {
+func (h *JSONProtocol) Dispatch(ctx context.Context, ci *event.Event) (*event.Event, error) {
```

Comment:
eventually I think this is probably the right thing to do (actually, kill dispatch is probably the right thing to do, the contract is whack imo even if the abstraction was useful while we needed it), and now that the code is all here we should probably proceed, I guess this seems like a lot of extra work to add here and test considering we're going to kill http/json in the relatively near future. am not sure it's worth undoing, the dispatch signature change does seem responsible for a large number of lines in this patch and I do think we'll kill it. previously I guess we were only writing back an http request, we could likely following dispatch translate that into an event (if it's not already), which would mean that there's only one code path that needs to do that translation (vs 2). again, not sure it's worth undoing, thinking out loud.